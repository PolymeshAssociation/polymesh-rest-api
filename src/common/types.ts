/* istanbul ignore file */

export interface Entity<Serialized> {
  uuid: string;

  toHuman(): Serialized;
}

// eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-explicit-any
export type Class<T extends {} = {}> = new (...args: any[]) => T;

export enum TransactionType {
  Single = 'single',
  Batch = 'batch',
}

export enum CalendarUnit {
  Second = 'Second',
  Minute = 'Minute',
  Hour = 'Hour',
  Day = 'Day',
  Week = 'Week',
  Month = 'Month',
  Year = 'Year',
}

/**
 * determines how transactions are processed
 */
export enum ProcessMode {
  /**
   * Sign and submit the transaction to the chain. Responds when transaction is in a finalized block
   */
  Submit = 'submit',
  /**
   * Sign and submit the transaction to the chain. Responds immediately, and posts status updates as the transaction is processed
   */
  SubmitWithCallback = 'submitWithCallback',
  /**
   * Return an unsigned transaction payload
   */
  Offline = 'offline',
  /**
   * Perform transaction validation, but does not perform the transaction
   */
  DryRun = 'dryRun',

  AMQP = 'AMQP',
}
