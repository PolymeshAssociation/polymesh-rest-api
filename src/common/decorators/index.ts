export * from '~/common/decorators/swagger';
export * from '~/common/decorators/transformation';
export * from '~/common/decorators/validation';
export * from '~/common/decorators/api-transaction-response';
