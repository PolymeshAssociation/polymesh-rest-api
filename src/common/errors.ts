/* istanbul ignore file */

export enum AppErrorCode {
  NotFound = 'NotFound',
  Conflict = 'Conflict',
  Config = 'Config',
  Validation = 'Validation',
  Unauthorized = 'Unauthorized',
  Unprocessable = 'Unprocessable',
  Internal = 'Internal',
}

export abstract class AppError extends Error {
  public readonly code: AppErrorCode;
}

export class AppNotFoundError extends AppError {
  public readonly code = AppErrorCode.NotFound;

  constructor(id: string, resource: string) {
    const identifierMessage = id !== '' ? `: with identifier: "${id}"` : '';

    super(`Not Found: ${resource} was not found${identifierMessage}`);
  }
}

export class AppConflictError extends AppError {
  public readonly code = AppErrorCode.Conflict;

  constructor(id: string, resource: string) {
    super(`Conflict: ${resource} already exists with unique identifier: "${id}"`);
  }
}

export class AppConfigError extends AppError {
  public readonly code = AppErrorCode.Config;

  constructor(key: string, message: string) {
    super(`Config: ${key}: ${message}`);
  }
}
export class AppValidationError extends AppError {
  public readonly code = AppErrorCode.Validation;

  constructor(message: string) {
    super(`Validation: ${message}`);
  }
}

export class AppUnauthorizedError extends AppError {
  public readonly code = AppErrorCode.Unauthorized;

  constructor(message: string) {
    super(`Unauthorized: ${message}`);
  }
}

export class AppUnprocessableError extends AppError {
  public readonly code = AppErrorCode.Unprocessable;

  constructor(message: string) {
    super(`Unprocessable: ${message}`);
  }
}

export class AppInternalError extends AppError {
  public readonly code = AppErrorCode.Internal;

  constructor(message: string) {
    super(`Internal: ${message}`);
  }
}

export function isAppError(err: Error): err is AppError {
  return err instanceof AppError;
}
