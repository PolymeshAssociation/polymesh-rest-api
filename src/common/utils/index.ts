import { ModuleName, TxTags } from '@polymeshassociation/polymesh-sdk/types';
import { randomBytes } from 'crypto';
import { flatten } from 'lodash';
import { promisify } from 'util';

import { TransactionBaseDto } from '~/common/dto/transaction-base-dto';
import { NotificationPayloadModel } from '~/common/models/notification-payload-model';
import { TransactionQueueModel } from '~/common/models/transaction-queue.model';
import { EventType } from '~/events/types';
import { NotificationPayload } from '~/notifications/types';
import { TransactionResult } from '~/transactions/transactions.util';

/* istanbul ignore next */
export function getTxTags(): string[] {
  return flatten(Object.values(TxTags).map(txTag => Object.values(txTag)));
}

/* istanbul ignore next */
export function getTxTagsWithModuleNames(): string[] {
  const txTags = getTxTags();
  const moduleNames = Object.values(ModuleName);
  return [...moduleNames, ...txTags];
}

export type TransactionResponseModel = NotificationPayloadModel | TransactionQueueModel;

/**
 * A helper type that lets a service return a QueueResult or a Subscription Receipt
 */
export type ServiceReturn<T> = Promise<
  NotificationPayload<EventType.TransactionUpdate> | TransactionResult<T>
>;

/**
 * A helper type that lets a controller return a Model or a Subscription Receipt if webhookUrl is being used
 */
export type TransactionResolver<T> = (
  res: TransactionResult<T>
) => Promise<TransactionQueueModel> | TransactionQueueModel;

/**
 * A helper function that transforms a service result for a controller. A controller can pass a resolver for a detailed return model, otherwise the transaction details will be used as a default
 */
export const handleServiceResult = <T>(
  result: NotificationPayloadModel | TransactionResult<T>,
  resolver: TransactionResolver<T> = basicModelResolver
): NotificationPayloadModel | Promise<TransactionQueueModel> | TransactionQueueModel => {
  if ('transactions' in result) {
    return resolver(result);
  }

  return new NotificationPayloadModel(result);
};

/**
 * A helper function for controllers when they should return a basic TransactionQueueModel
 */
const basicModelResolver: TransactionResolver<unknown> = ({ transactions, details }) => {
  return new TransactionQueueModel({ transactions, details });
};

/**
 * Generate base64 encoded, cryptographically random bytes
 *
 * @note random byte length given, not the encoded string length
 */
export const generateBase64Secret = async (byteLength: number): Promise<string> => {
  const buf = await promisify(randomBytes)(byteLength);

  return buf.toString('base64');
};

/**
 * Helper class to ensure a code path is unreachable. For example this can be used for ensuring switch statements are exhaustive
 */
export class UnreachableCaseError extends Error {
  /** This should never be called */
  constructor(val: never) {
    super(`Unreachable case: ${JSON.stringify(val)}`);
  }
}

export const extractTxBase = <T extends TransactionBaseDto>(
  params: T
): {
  base: TransactionBaseDto;
  args: Omit<T, keyof TransactionBaseDto>;
} => {
  const { signer, webhookUrl, dryRun, ...args } = params;
  return {
    base: { signer, webhookUrl, dryRun },
    args,
  };
};
