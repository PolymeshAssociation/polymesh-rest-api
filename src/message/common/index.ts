/* eslint-disable no-restricted-imports */
export * from './message.service';
export * from './types';
