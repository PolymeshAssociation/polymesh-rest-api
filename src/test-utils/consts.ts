import { createMock } from '@golevelup/ts-jest';
import { BigNumber } from '@polymeshassociation/polymesh-sdk';
import { Account , PayingAccountType, TransactionStatus } from '@polymeshassociation/polymesh-sdk/types';

import { UserModel } from '~/users/model/user.model';

const signer = 'alice';
const did = '0x01'.padEnd(66, '0');
const dryRun = false;

const user = new UserModel({
  id: '-1',
  name: 'TestUser',
});

const resource = {
  type: 'TestResource',
  id: '-1',
} as const;

export const testAccount = createMock<Account>({ address: 'address' });
export const txResult = {
  transactions: ['transaction'],
  details: {
    status: TransactionStatus.Succeeded,
    fees: {
      gas: new BigNumber(1),
      protocol: new BigNumber(1),
      total: new BigNumber(1),
    },
    supportsSubsidy: false,
    payingAccount: {
      address: did,
      balance: new BigNumber(1),
      type: PayingAccountType.Caller,
    },
  },
};

export const testValues = {
  signer,
  did,
  user,
  resource,
  testAccount,
  txResult,
  dryRun,
};
