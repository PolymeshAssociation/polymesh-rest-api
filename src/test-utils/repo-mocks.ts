/* istanbul ignore file */

import { createMock } from '@golevelup/ts-jest';
import { ValueProvider } from '@nestjs/common';

import { ApiKeyRepo } from '~/auth/repos/api-key.repo';
import { UsersRepo } from '~/users/repo/user.repo';

export const mockApiKeyRepoProvider: ValueProvider<ApiKeyRepo> = {
  provide: ApiKeyRepo,
  useValue: createMock<ApiKeyRepo>(),
};

export const mockUserRepoProvider: ValueProvider<UsersRepo> = {
  provide: UsersRepo,
  useValue: createMock<UsersRepo>(),
};

export class MockPostgresApiRepository {
  create = jest.fn();
  findOneBy = jest.fn();
  delete = jest.fn();
  save = jest.fn();
}

export class MockPostgresUserRepository {
  findOneBy = jest.fn();
  save = jest.fn();
  delete = jest.fn();
  create = jest.fn();
}
