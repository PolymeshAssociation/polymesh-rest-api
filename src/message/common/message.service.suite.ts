/* istanbul ignore file */

import { AddressName, QueueName } from '~/common/utils/amqp';
import { MessageService } from '~/message/common/message.service';
import { AnyModel } from '~/offline-recorder/model/any.model';

export const testMessageService = async (service: MessageService): Promise<void> => {
  describe('sending and receiving messages', () => {
    it('should register a listener', async () => {
      const mockListener = jest.fn();
      await service.registerListener(QueueName.Requests, mockListener, AnyModel);

      await service.sendMessage(AddressName.Requests, { someKey: 'someValue' });

      expect(mockListener).toHaveBeenCalled();
    });
  });
};
