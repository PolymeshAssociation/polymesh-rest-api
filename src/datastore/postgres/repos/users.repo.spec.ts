import { when } from 'jest-when';
import { Repository, TypeORMError } from 'typeorm';

import { AppConflictError } from '~/common/errors';
import { User } from '~/datastore/postgres/entities/user.entity';
import { PostgresUsersRepo } from '~/datastore/postgres/repos/users.repo';
import { testUser } from '~/test-utils/consts';
import { MockPostgresUserRepository } from '~/test-utils/repo-mocks';
import { UsersRepo } from '~/users/repo/user.repo';

const uniqueViolation = new TypeORMError('duplicate key value violates unique constraint');

describe(`PostgresUsersRepo does not meet ${UsersRepo.type} requirements`, () => {
  const mockRepository = new MockPostgresUserRepository();
  const repo = new PostgresUsersRepo(mockRepository as unknown as Repository<User>);
  let _id = 1;

  mockRepository.create.mockImplementation(params => params);
  mockRepository.findOneBy.mockResolvedValue(null);

  mockRepository.save.mockImplementation(async user => {
    const { name } = user;
    user.id = _id++;
    when(mockRepository.save)
      .calledWith(expect.objectContaining({ name }))
      .mockRejectedValue(uniqueViolation);
    when(mockRepository.findOneBy).calledWith({ name }).mockResolvedValue(user);
  });

  UsersRepo.test(repo);
});

describe('PostgresApiKeyRepo', () => {
  const mockRepository = new MockPostgresUserRepository();
  const repo = new PostgresUsersRepo(mockRepository as unknown as Repository<User>);
  const name = testUser.name;
  describe('method: createUser', () => {
    it('should transform TypeORM unique violation into AppConflictError', () => {
      mockRepository.save.mockRejectedValue(uniqueViolation);

      return expect(repo.createUser({ name })).rejects.toThrow(AppConflictError);
    });

    it('should not transform generic TypeORM errors', () => {
      const typeOrmError = new TypeORMError('Test TypeORM error');

      mockRepository.save.mockRejectedValue(typeOrmError);

      return expect(repo.createUser({ name })).rejects.toThrowError(typeOrmError);
    });

    it('should throw errors as they are', () => {
      const error = new Error('Testing for when something goes wrong');
      mockRepository.save.mockRejectedValue(error);

      return expect(repo.createUser({ name })).rejects.toThrowError(error);
    });
  });
});
