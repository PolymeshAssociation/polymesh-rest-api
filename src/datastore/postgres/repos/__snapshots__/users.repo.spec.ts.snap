// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostgresUsersRepo does not meet User requirements method: createUser should create a user 1`] = `
UserModel {
  "id": "1",
  "name": "Alice",
}
`;

exports[`PostgresUsersRepo does not meet User requirements method: findByName should find the created user 1`] = `
UserModel {
  "id": "1",
  "name": "Alice",
}
`;
